File: src/0092.reverse-linked-list-ii/main.c
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be
// found in the LICENSE file.

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
  int val;
  struct ListNode* next;
};

static struct ListNode* listNew(int val) {
  struct ListNode* list = malloc(sizeof(struct ListNode));
  assert(list != NULL);
  list->val = val;
  list->next = NULL;
  return list;
}

static struct ListNode* listFromArray(int arr[], int len) {
  assert(len >= 0);
  struct ListNode* list = NULL;
  for (int i = len - 1; i >= 0; --i) {
    struct ListNode* node = listNew(arr[i]);
    node->next = list;
    list = node;
  }
  return list;
}

static void listFree(struct ListNode* list) {
  assert(list != NULL);
  while (list != NULL) {
    struct ListNode* node = list;
    list = list->next;
    free(node);
  }
}

static struct ListNode* reverseBetween(struct ListNode* head, int left, int right) {
  assert(head != NULL);
  assert(left >= 1 && left <= right);
  if (left == right) {
    return head;
  }
  left -= 1;
  right -= 1;

  struct ListNode* range_prev = NULL;
  struct ListNode* range_start = NULL;
  struct ListNode* range_end = NULL;
  struct ListNode* curr = head;
  struct ListNode* prev = NULL;
  struct ListNode* next = NULL;

  int index = 0;
  while (index < left) {
    range_prev = curr;
    curr = curr->next;
    index += 1;
  }

  range_start = curr;

  while (index < right) {
    // switch direction
    next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
    index += 1;
  }
  range_end = curr;
  if (range_prev == NULL) {
    head = range_end;
  } else {
    range_prev->next = range_end;
  }
  range_start->next = range_end->next;
  range_end->next = prev;

  return head;
}

static void checkSolution(int arr[], int len, int left, int right) {
  struct ListNode* list = listFromArray(arr, len);
  struct ListNode* reversed = reverseBetween(list, left, right);
  list = reversed;
  while (reversed != NULL) {
    printf("%d, ", reversed->val);
    reversed = reversed->next;
  }
  printf("\n");
  listFree(list);
}

int main() {
  int arr1[] = {1, 2, 3, 4, 5}; 
  checkSolution(arr1, sizeof(arr1) / sizeof(arr1[0]), 2, 4);

  int arr2[] = {5}; 
  checkSolution(arr2, sizeof(arr2) / sizeof(arr2[0]), 1, 1);

  int arr3[] = {3, 5}; 
  checkSolution(arr3, sizeof(arr3) / sizeof(arr3[0]), 1, 2);

  return 0;
}

File: src/0092.reverse-linked-list-ii/src/main.rs
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be found
// in the LICENSE file.

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[must_use]
    #[inline]
    pub const fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }

    #[must_use]
    pub fn from_slice(slice: &[i32]) -> Option<Box<ListNode>> {
        let mut list = None;
        for &val in slice.iter().rev() {
            list = Some(Box::new(ListNode { val, next: list }));
        }
        list
    }

    pub fn debug_print(head: &Option<Box<ListNode>>) {
        let mut node_ref = head;
        print!("head: [");
        while node_ref.is_some() {
            let val: i32 = node_ref.as_ref().unwrap().val;
            print!("{val}, ");
            node_ref = &node_ref.as_ref().unwrap().next;
        }
        println!("]");
    }
}

pub fn reverse_between1(
    head: Option<Box<ListNode>>,
    left: i32,
    right: i32,
) -> Option<Box<ListNode>> {
    // 反转链表里面实现的, 只反转前面几个元素.
    fn reverse_list(head: Option<Box<ListNode>>, len: i32) -> Option<Box<ListNode>> {
        debug_assert!(len > 0);
        let mut current_node = None;
        let mut next_node = head;

        for _ in 0..len {
            let new_next_node = next_node.as_mut()?.next.take();
            // 反转一个节点
            next_node.as_mut()?.next = current_node;
            current_node = next_node;
            next_node = new_next_node;
        }

        // 遍历反转后的链表, 重新连接尾部元素.
        let mut node_ref = &mut current_node;
        for _ in 0..(len - 1) {
            node_ref = &mut node_ref.as_mut().unwrap().next;
        }
        node_ref.as_mut().unwrap().next = next_node;

        // 最后返回新的 list head.
        current_node
    }

    debug_assert!(left > 0 && left <= right);
    // 处理边角情况
    if left == right {
        return head;
    }

    // 添加一个dummy header, 方便反转第一个节点.
    let mut dummy_head = Some(Box::new(ListNode { val: 0, next: head }));
    let mut current_ref = &mut dummy_head;
    let mut index: i32 = 0;

    while current_ref.is_some() {
        index += 1;
        // 开始反转
        if index == left {
            // 首先把 current->next 取出来
            let rev_head = current_ref.as_mut()?.next.take();
            let new_head = reverse_list(rev_head, right - left + 1);
            // 终止反转
            // 进行新链表的拼装.
            current_ref.as_mut()?.next = new_head;
        }
        current_ref = &mut current_ref.as_mut()?.next;
    }

    dummy_head.as_mut()?.next.take()
}

pub type SolutionFn = fn(Option<Box<ListNode>>, i32, i32) -> Option<Box<ListNode>>;

fn check_solution(func: SolutionFn) {
    let head = ListNode::from_slice(&[1, 2, 3, 4, 5]);
    ListNode::debug_print(&head);
    let expected = ListNode::from_slice(&[1, 4, 3, 2, 5]);
    let result = func(head, 2, 4);
    ListNode::debug_print(&result);
    assert_eq!(result, expected);

    let head = ListNode::from_slice(&[5]);
    let expected = ListNode::from_slice(&[5]);
    let result = func(head, 1, 1);
    ListNode::debug_print(&result);
    assert_eq!(result, expected);
}

fn main() {
    check_solution(reverse_between1);
}

#[cfg(test)]
mod tests {
    use super::{check_solution, reverse_between1};

    #[test]
    fn test_reverse_between1() {
        check_solution(reverse_between1);
    }
}
